ci(miri): infra-fast job hangs at 60m timeout + infra-slow budget knife-edge

## Problem

In scheduled run #31983525568 (`ci.yml`, 2026-08-17) two infra Miri jobs died at exactly the 60-minute timeout — this is **not** the boring-sys2 FFI issue (tracked in #750):

- **Miri (infra-fast):** last successful test at `01:11:28Z` (`infrastructure::output::frontmatter::tests::test_generate_without_rich_metadata_backward_compatible`), then 47 minutes of complete silence — one test hung — killed at `01:58:00Z`. It has been cancelled at the timeout in every run since the job was split (c4857ed, Aug 7).
- **Miri (infra-slow):** healthy but slow — 286 tests completed, still inside `infrastructure::obsidian::metadata` when killed at the 60-minute mark. On Aug 10 it managed to finish; budget is on a knife edge.

## Evidence

From the run log (`gh run view 31983525568 --log`):
```
Miri (infra-fast) ... test frontmatter::tests::test_generate_without_rich_metadata_backward_compatible ... ok   # 01:11:28Z
Miri (infra-fast) ... ##[error]The operation was canceled.                                                       # 01:58:00Z
```
The hang is inside the infra-fast module set (`cpu_pool`, `stream`, `error`, `user_agent`, `bridge`, `output`, `frontmatter`... alphabetical order within the filter), after the `output::frontmatter` block. Known repo pattern: tokio time-driver does not advance under Miri (`retry_policy.rs:120`, `resource_downloader.rs:480-780`), and multi-thread `spawn_blocking` hangs (`bridge.rs:261-400`).

## Work needed

1. Reproduce locally: `MIRIFLAGS="-Zmiri-tree-borrows -Zmiri-disable-isolation -Zmiri-ignore-leaks" cargo miri test --lib -- infrastructure::output:: infrastructure::autotuning:: infrastructure::content_processing:: infrastructure::scraper:: ...` to identify the hanging test.
2. Gate it with `#[cfg(not(miri))]` + comment, or split the fast/slow module set so each fits its 60-minute budget.
3. Consider re-balancing the infra-slow filter (286 tests is marginal at 60m).

This is a hang-root-cause investigation, not a one-liner — hence split from #750.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -565,11 +565,17 @@ jobs:
   # ============================================================================
   # MIRI INFRA-SLOW: crawler/obsidian/export/http/network/converter + pure-Rust
   # downloader submodules. 60 min timeout for the same reason as infra-fast.
-  # wikilinks/syntax_highlight explicitly skipped — FFI-heavy, can't execute
-  # under Miri. downloader: only pure-Rust submodules (cookie_bridge/spa_detector)
+  # syntax_highlight is excluded at source via #[cfg(all(test, not(miri)))];
+  # wikilinks (pure pulldown_cmark, no FFI) runs and passes under Miri. The
+  # trailing `--skip wikilinks`/`--skip syntax_highlight` flags are inert
+  # (parsed after the `--` positional filters) — kept as documentation.
+  # downloader: only pure-Rust submodules (cookie_bridge/spa_detector)
   # — FFI (wreq, obscura, chromiumoxide) and sysinfo-dependent (hybrid_router,
   # resource_governor) excluded via --skip / #[cfg(not(miri))] (#515).
   # resource_governor is covered by TSan (sanitizers.yml), where sysconf works.
+  # Budget (#751): memory_manager's test_handle_disk_swapping_with_swap is
+  # #[cfg_attr(miri, ignore)] — it alone burned ~27 min wall, leaving the rest
+  # of this job at ~30–33 min of the 60 min timeout (knife-edge resolved).
   # ============================================================================
   miri-infra-slow:
     name: Miri (infra-slow)

--- a/crates/webfang_core/src/infrastructure/crawler/memory_manager.rs
+++ b/crates/webfang_core/src/infrastructure/crawler/memory_manager.rs
@@ -185,6 +185,7 @@ mod tests {
         assert!(result.is_ok());
     }
 
+    #[cfg_attr(miri, ignore)] // 15,000-URL disk write takes ~27m wall under Miri (#751, run 31983525568: 01:07→01:35)
     #[test]
     fn test_handle_disk_swapping_with_swap() {
         let temp_dir = TempDir::new().unwrap();

--- a/crates/webfang_core/src/infrastructure/scraper/dom_inspector.rs
+++ b/crates/webfang_core/src/infrastructure/scraper/dom_inspector.rs
@@ -248,6 +248,19 @@ mod tests {
         let suggestions = inspector.suggest(&document, ".nonexistent");
         assert!(suggestions.is_empty());
     }
+}
+
+// DefaultDomInspector tests construct and drop `scraper::Selector` values (via
+// `Selector::parse` in `collect_candidates`/`inspect`/`suggest`). Dropping a
+// selector trips a Tree Borrows UB inside servo_arc's `ArcUnion`
+// (`servo_arc` 0.4.3 lib.rs:535, `Arc::drop`) — the same defect that gated
+// readability/author_extractor (5a2adc2, #515) — and large DOMs
+// (`builder_methods`, 600 divs) spin indefinitely under the interpreter. In CI
+// this module hung the whole infra-fast job until the 60m timeout (#751, run
+// 31983525568: 31 tests never completed). Regular (non-Miri) runs keep coverage.
+#[cfg(all(test, not(miri)))]
+mod default_dom_inspector_tests {
+    use super::*;
 
     // --- DefaultDomInspector::inspect tests ---
 

